### PR TITLE
bpo-44708: Only re-run test methods that match names of previously failing test methods

### PR DESCRIPTION
**Pull request by ambv**
Details on the issue. It's probably easiest to review this commit by commit as every commit is an atomic change.

I'd also like to backport to 3.10 and 3.9 to speed up open PRs on those branches as well.

https://bugs.python.org/issue44708

### CODE REVIEW

File: Lib/test/libregrtest/main.py [hunk 1]
```diff
@@ -99,34 +99,32 @@ def get_executed(self):
...
 
-        if ok not in (CHILD_ERROR, INTERRUPTED) and not rerun:
-            self.test_times.append((result.test_time, test_name))
+        if not isinstance(result, (ChildError, Interrupted)) and not rerun:
+            self.test_times.append((result.duration_sec, test_name))
 
-        if ok == PASSED:
+        if isinstance(result, Passed):
             self.good.append(test_name)
-        elif ok in (FAILED, CHILD_ERROR):
```
Comment (ambv) [inline Lib/test/libregrtest/main.py:110]: The different order here is required due to how inheritance works. `isinstance(Failed)` would catch all the lower cases.

File: Lib/test/libregrtest/main.py [hunk 2]
```diff
@@ -99,34 +99,32 @@ def get_executed(self):
...
+        elif isinstance(result, Interrupted):
             self.interrupted = True
-        elif ok == TIMEOUT:
-            self.bad.append(test_name)
         else:
-            raise ValueError("invalid test result: %r" % ok)
+            raise ValueError("invalid test result: %r" % result)
 
-        if rerun and ok not in {FAILED, CHILD_ERROR, INTERRUPTED}:
+        if rerun and not isinstance(result, (Failed, Interrupted)):
```
Comment (ambv) [inline Lib/test/libregrtest/main.py:127]: `ChildError` is a `Failed` subclass so we don't have to enumerate it. Additionally, I feel like this list was missing other failures like uncaught exceptions, env changeds, and refleaks. Now it doesn't.

File: Lib/test/libregrtest/main.py [hunk 3]
```diff
@@ -314,15 +312,31 @@ def rerun_failed_tests(self):
...
+            failures = result.failures or []
+            error_names = [e[0].split(" ")[0] for e in errors]
+            failure_names = [f[0].split(" ")[0] for f in failures]
             self.ns.verbose = True
+            orig_match_tests = self.ns.match_tests
+            if errors or failures:
+                if self.ns.match_tests is None:
+                    self.ns.match_tests = []
+                self.ns.match_tests.extend(error_names)
+                self.ns.match_tests.extend(failure_names)
```
Comment (ambv) [inline Lib/test/libregrtest/main.py:329]: This is the money shot.

File: Lib/test/libregrtest/runtest.py [hunk 4]
```diff
@@ -12,33 +11,109 @@
...
-# Test result constants.
-PASSED = 1
-FAILED = 0
-ENV_CHANGED = -1
-SKIPPED = -2
-RESOURCE_DENIED = -3
-INTERRUPTED = -4
-CHILD_ERROR = -5   # error in a child process
-TEST_DID_NOT_RUN = -6
-TIMEOUT = -7
```
Comment (ambv) [inline Lib/test/libregrtest/runtest.py:29]: Replacing those with objects was necessary to store related context when returning, which allows matching test method names in `main.py`. This helps with consistency when running across processes as well.

File: Lib/test/libregrtest/runtest.py [hunk 5]
```diff
@@ -156,36 +217,26 @@ def _runtest(ns, test_name):
...
-    Returns the tuple (result, test_time, xml_data), where result is one
-    of the constants:
-
-        INTERRUPTED      KeyboardInterrupt
-        RESOURCE_DENIED  test skipped because resource denied
-        SKIPPED          test skipped for some other reason
-        ENV_CHANGED      test failed because it changed the execution environment
-        FAILED           test failed
-        PASSED           test passed
-        EMPTY_TEST_SUITE test ran no subtests.
```
Comment (ambv) [inline Lib/test/libregrtest/runtest.py:187]: This docstring was outdated anyway.

File: Lib/test/libregrtest/runtest_mp.py [hunk 6]
```diff
@@ -36,21 +38,21 @@
...
     if ns.failfast and is_failed(result, ns):
         return True
     return False
 
 
-def parse_worker_args(worker_args):
+def parse_worker_args(worker_args) -> tuple[Namespace, str]:
     ns_dict, test_name = json.loads(worker_args)
-    ns = types.SimpleNamespace(**ns_dict)
+    ns = Namespace(**ns_dict)
```
Comment (ambv) [inline Lib/test/libregrtest/runtest_mp.py:51]: Custom Namespace subclass pays off.

File: Lib/test/libregrtest/runtest.py [hunk 7]
```diff
@@ -283,37 +336,43 @@ def _runtest_inner(ns, test_name, display_failure=True):
...
         if not ns.quiet and not ns.pgo:
             print(f"{test_name} skipped -- {msg}", flush=True)
-        return SKIPPED
+        return Skipped(test_name)
+    except support.TestFailedWithDetails as exc:
+        msg = f"test {test_name} failed"
+        if display_failure:
+            msg = f"{msg} -- {exc}"
+        print(msg, file=sys.stderr, flush=True)
+        return Failed(test_name, errors=exc.errors, failures=exc.failures)
```
Comment (ambv) [inline Lib/test/libregrtest/runtest.py:349]: This is how we store the failed and erroring method names to be matched later for re-running.

File: Lib/test/libregrtest/runtest.py [hunk 8]
```diff
@@ -12,33 +11,109 @@
...
-}
+from test.libregrtest.utils import clear_caches, format_duration, print_warning
+
+
+class TestResult:
+    def __init__(
+        self,
+        name: str,
+        duration_sec: float = 0.0,
+        xml_data: list[str] | None = None,
```
Comment (Fidget-Spinner) [inline Lib/test/libregrtest/runtest.py:24]: Drive-by comment: ``T1 | T2`` syntax is 3.10 only and won't be able to land in 3.9 unless you use ``from __future__ import annotations`` or stringify them manually. Ditto for other occurrences below.

I don't have enough experience in `libregrtest` to review the rest, but this is a change I look forward to :).

File: Lib/test/libregrtest/main.py (hunk 2 above)
Comment (pablogsal) [reply Lib/test/libregrtest/main.py:127]: For refleaks/envchanged I don't know if it makes sense to rerun because it those fail, then there is no way that the test suite is going to succeed after that, no? OTOH I don't see how it would hurt (just let's make sure that it works as we expect and it doesn't interfere with the refleak machinery).

File: Lib/test/libregrtest/runtest.py (hunk 4 above)
Comment (pablogsal) [reply Lib/test/libregrtest/runtest.py:29]: This is an excellent change, dealing with these have been quite painful before

File: Lib/test/libregrtest/runtest.py [hunk 9]
```diff
@@ -17,28 +17,104 @@
...
-    FAILED: '%s failed',
-    ENV_CHANGED: '%s failed (env changed)',
-    SKIPPED: '%s skipped',
-    RESOURCE_DENIED: '%s skipped (resource denied)',
-    INTERRUPTED: '%s interrupted',
-    CHILD_ERROR: '%s crashed',
-    TEST_DID_NOT_RUN: '%s run no tests',
-    TIMEOUT: '%s timed out',
-}
+class TestResult:
```
Comment (pablogsal) [inline Lib/test/libregrtest/runtest.py:19]: Nit (feel free to ignore): you can probably make a dataclass out of these classes

File: Lib/test/libregrtest/runtest.py [hunk 10]
```diff
@@ -284,34 +337,40 @@ def _runtest_inner(ns: Namespace, test_name: str, display_failure: bool = True)
...
             print(f"{test_name} skipped -- {msg}", flush=True)
-        return SKIPPED
+        return Skipped(test_name)
+    except support.TestFailedWithDetails as exc:
+        msg = f"test {test_name} failed"
+        if display_failure:
+            msg = f"{msg} -- {exc}"
+        print(msg, file=sys.stderr, flush=True)
+        return Failed(test_name, errors=exc.errors, failures=exc.failures)
     except support.TestFailed as exc:
```
Comment (pablogsal) [inline Lib/test/libregrtest/runtest.py:350]: In what scenarios are we are raising `Failed` but not `TestFailedWithDetails` now?

Comment (ambv) [reply Lib/test/libregrtest/runtest.py:350]: There are quite a few bare TestFailed in custom `test_main` functions in various `test_*.py` files. Example:

https://github.com/python/cpython/blob/c05a790693b3b00ac7cb5b96ad416ca9d8ecb14a/Lib/test/test_ssl.py#L5044

File: Lib/test/libregrtest/main.py [hunk 11]
```diff
@@ -311,10 +312,24 @@ def rerun_failed_tests(self):
...
-        for test_name in self.rerun:
-            self.log(f"Re-running {test_name} in verbose mode")
+        rerun_list = self.rerun[:]
+        self.rerun = []
+        for result in rerun_list:
+            test_name = result.name
+            errors = result.errors or []
+            failures = result.failures or []
+            error_names = [e[0].split(" ")[0] for e in errors]
+            failure_names = [f[0].split(" ")[0] for f in failures]
```
Comment (pablogsal) [inline Lib/test/libregrtest/main.py]: I had some trouble following what `e[0].split(" ")[0]` is doing (as in, what `e[0]` actually is). What about:


```suggestion
            error_names = [test_full_name.split(" ")[0] for (test_full_name, *_) in errors]
            failure_names = [test_full_name.split(" ")[0] for (test_full_name, *_) in failures]
```

Comment (ambv) [reply Lib/test/libregrtest/main.py]: `test_name` in often used in libregrtest to denote the name of the test file. I'll use `method_name` instead.

File: Lib/test/libregrtest/main.py [hunk 12]
```diff
@@ -311,10 +312,24 @@ def rerun_failed_tests(self):
...
             self.ns.verbose = True
+            if errors or failures:
+                if self.ns.match_tests is None:
+                    self.ns.match_tests = []
+                self.ns.match_tests.extend(error_names)
```
Comment (pablogsal) [inline Lib/test/libregrtest/main.py:328]: Question: Do we want to rerun errors? In theory those are things like failures to set up the test and similar. Do you think that this is something we should run again?

**Review by pablogsal — APPROVED** [review]
Fantastic job! 🚀  In general LGTM!

I left some comments with some questions and suggestions.

File: Lib/test/libregrtest/main.py (hunk 12 above)
Comment (ambv) [reply Lib/test/libregrtest/main.py:328]: The only case where we gather errors is with `TestFailedWithDetails` which means those are `unittest` failures and/or errors. We've been re-running the entire file in this scenario before. I think we should still as some of those are exceptions raised due to race conditions. We might think of tightening this up in a future update.

File: Lib/test/libregrtest/runtest_mp.py (hunk 6 above)
Comment (pablogsal) [reply Lib/test/libregrtest/runtest_mp.py:51]: 👏

File: Lib/test/libregrtest/main.py (hunk 2 above)
Comment (ambv) [reply Lib/test/libregrtest/main.py:127]: This function is a little cryptic. The `rerun` boolean here means "we are already re-running" and this if statement figures out whether we should remove the current test that just re-ran from "bad" tests (which were populated on the first run). The bug that was there previously is that a refleak or env change in the second run would count as an "ok" outcome and the test would be removed from "bad", leading to a green CI badge that's wrong.
